File: java/ReverseLinkedList.java
/**
 * Author: Chih-Jye Wang
 * Date  : Oct 11, 2015
 *
 * Reverse a singly linked list.
 *
 * Hint:
 * A linked list can be reversed either iteratively or recursively. Could you implement both?
 */

public class ReverseLinkedList {
    public ListNode reverseList(ListNode head) {
        ListNode r = null, p = head;
        while(p != null) {
            ListNode t = p;
            p = p.next;
            t.next = r;
            r = t;
        }
        return r;
    }
}

File: ruby/reverse_linked_list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
    res = nil
    while head
        node = head
        head = head.next
        if res
            node.next = res
            res = node
        else
            node.next = nil
            res = node
        end
    end
    return res
end
